c10/core/TensorImpl.cpp

```diff
@@ -85,6 +85,40 @@ bool TensorImpl::compute_contiguous() const {
   return is_contiguous;
 }
 
+bool TensorImpl::compute_channels_last_contiguous() const {
+  if (dim() == 4) {
+    int64_t expected = 1;
+    for (auto& d : {1, 3, 2, 0}) {
+      if (size(d) != 1) {
```

Comment:
why skip for size 1? better to be safer and check it.

Also, there might be special effects if H=W=1 but I guess it's less likely.

Reply:
Because size 1 strides is broken in PyTorch, for example:

```python
N = 10; C = 3; H = 32; W = 16;                                                                                                             
x = torch.randn(N,H,W,C)                                                                                                                     
x = x.permute(0,3,1,2)                                                                                                                       
y = x[0]                                                                                                                                     
z = y.unsqueeze(0)    
y.stride()                                                                                                                                   
Out[40]: (1, 48, 3)

In [41]: z.stride()                                                                                                                                   
Out[41]: (3, 1, 48, 3)
```

Reply:
I don't follow the reasoning here. There doesn't seem to be anything wrong with `x[0]`, so you must be complaining about the behavior of `unsqueeze`. But the stride of a size one dimension doesn't actually matter, since it doesn't ever get used. Maybe the choice of the last element is a bit goofy; we probably did it because it's compatible with Numpy's behavior:

```
>>> N = 10; C = 3; H = 32; W = 16
>>> x = np.zeros((N,H,W,C))
>>> x = x.transpose(0,3,1,2)
>>> y = x[0]
>>> y.strides
(8, 384, 24)
>>> np.expand_dims(y, 0).strides
(24, 8, 384, 24)
```

Reply:
Maybe your claim is that, since strides don't matter for size one dimensions, we should ignore them. I can get behind that reason. But I'm not sure I would jump to the conclusion that it is "broken".

Reply:
In any case this should be amply documented :>

Reply:
Hmm, you should probably also skip zero size dimensions, following this reasoning.

Reply:
Agree, _broken_ is incorrect word here, what I meant is they carry zero information.